pageserver/src/walreceiver.rs

```diff
@@ -1,170 +1,567 @@
-    let receivers = WAL_RECEIVERS.lock().unwrap();
+struct TimelineWalReceiverHandles {
+    broker_join_handle: JoinHandle<anyhow::Result<()>>,
+    cancellation_sender: watch::Sender<()>,
+}
 
-    receivers
-        .get(&(tenantid, timelineid))
-        .unwrap()
-        .wal_producer_connstr
-        .clone()
+impl TimelineWalReceiverHandles {
+    async fn shutdown(self, id: ZTenantTimelineId) -> anyhow::Result<()> {
+        self.cancellation_sender.send(()).context(
+            "Unexpected: cancellation sender is dropped before the receiver in the loop is",
+        )?;
+        debug!("Waiting for wal receiver for timeline {id} to finish");
+        self.broker_join_handle
+            .await
+            .with_context(|| format!("Failed to join the wal reveiver broker for timeline {id}"))?
+            .with_context(|| format!("Wal reveiver broker for timeline {id} failed to finish"))
+    }
 }
 
-//
-// This is the entry point for the WAL receiver thread.
-//
-fn thread_main(conf: &'static PageServerConf, tenant_id: ZTenantId, timeline_id: ZTimelineId) {
-    let _enter = info_span!("WAL receiver", timeline = %timeline_id, tenant = %tenant_id).entered();
-    info!("WAL receiver thread started");
+async fn start_timeline_wal_broker(
+    broker_prefix: &str,
+    etcd_endpoints: &[Url],
+    id: ZTenantTimelineId,
+    timeline: &Arc<DatadirTimelineImpl>,
+    connect_timeout: Duration,
+) -> anyhow::Result<TimelineWalReceiverHandles> {
+    let mut etcd_client = etcd_client::Client::connect(etcd_endpoints, None)
+        .await
+        .context("Failed to connect to etcd")?;
+    let (cancellation_sender, mut cancellation_receiver) = watch::channel(());
+
+    let mut subscription = etcd_broker::subscribe_to_safekeeper_timeline_updates(
+        &mut etcd_client,
+        SkTimelineSubscriptionKind::timeline(broker_prefix.to_owned(), id),
+    )
+    .await
+    .with_context(|| format!("Failed to subscribe for timeline {id} updates in etcd"))?;
+
+    let timeline = Arc::clone(timeline);
+    let broker_join_handle = tokio::spawn(async move {
+        info!("WAL receiver broker thread had started");
+
+        let mut wal_receiver_connection = None::<WalReceiverConnection>;
+        let (walreceiver_task_completion_sender, mut walreceiver_task_completion) =
+            watch::channel(Ok(()));
+        let walreceiver_task_completion_sender = Arc::new(walreceiver_task_completion_sender);
+        let mut connection_attempt = 0;
+
+        loop {
+            select! {
+                // TODO kb add here the walreceiver and the rest
+
+                // check for shutdown first
+                biased;
+                _ = cancellation_receiver.changed() => {
+                    info!("Loop cancelled, shutting down");
+                    break;
+                }
+                walreceiver_result = walreceiver_task_completion.changed() => {
```

Comment:
Looks like we have to keep the channel, since it's the only way for us to understand, whether the WAL receiver thread/task got any updates from safekeeper: it's crucial for us to detect the lagging ones and not sure there'a a way around it.

We also have to share some public info about WAL streaming via HTTP API now, so not sure the `join` approach works now.